relay-server/src/actors/store.rs

```diff
@@ -151,20 +156,143 @@ impl StoreForwarder {
+                        session_id: Uuid::nil(),
+                        distinct_id: Uuid::nil(),
+                        quantity: 1,
+                        seq: 0,
+                        received: protocol::datetime_to_timestamp(chrono::Utc::now()),
+                        started: 0f64,
+                        duration: None,
+                        errors: 0,
+                        release,
+                        environment,
+                        user_agent: user_agent.clone(),
+                        os,
+                        runtime,
+                        sdk: client.map(str::to_owned),
+                        retention_days: event_retention,
+                        status: "exited",
+                    };
+
+                    if aggregates.len() > MAX_EXPLODED_SESSIONS {
+                        log::warn!("aggregated session items exceed threshold");
+                    }
+
+                    for item in aggregates.into_iter().take(MAX_EXPLODED_SESSIONS) {
+                        message.started = protocol::datetime_to_timestamp(item.started);
+                        message.distinct_id = item
+                            .distinct_id
+                            .as_deref()
+                            .map(make_distinct_id)
+                            .unwrap_or_default();
+                        let overrode_ua = if let Some(ua) = item.user_agent {
+                            message.user_agent = Some(ua);
+                            true
+                        } else {
+                            false
+                        };
+                        if item.exited > 0 {
+                            message.errors = 0;
+                            message.status = "exited";
+                            message.quantity = item.exited;
+                            self.send_session_message(message.clone())?;
+                        }
+                        if item.errored > 0 {
+                            message.errors = 1;
+                            message.status = "errored";
+                            message.quantity = item.errored;
+                            self.send_session_message(message.clone())?;
+                        }
+                        if item.abnormal > 0 {
+                            message.errors = 1;
+                            message.status = "abnormal";
+                            message.quantity = item.abnormal;
+                            self.send_session_message(message.clone())?;
+                        }
+                        if item.crashed > 0 {
+                            message.errors = 1;
+                            message.status = "crashed";
+                            message.quantity = item.crashed;
+                            self.send_session_message(message.clone())?;
+                        }
+                        if overrode_ua {
+                            message.user_agent = user_agent.clone();
```

Comment:
Instead of mutating a shared session struct, it would be safer to have an immutable "prototype", clone that into your scope, mutate the cloned instance, and then pass it into `send_session_message`. Otherwise, it is too easy to forget resetting another attribute.